### PR TITLE
Use puppeteer 5

### PR DESCRIPTION
**Pull request by xtrime-ru**
Its latest version supported in browserless: https://hub.docker.com/r/browserless/chrome/tags?page=1&name=puppeteer-5

We checked this on multiple sites. All works fine.  New chrome versions have reduced CPU consumption.

### CODE REVIEW

**Comment by spekulatius** [comment]
Hello @xtrime-ru,

Thank you for your help with the upgrade! It's something @nesk and I got on the radar too! We are working on PuPHPeteer on te side :)

To allow accessing and identifying the resources on the PHP-side it would be good to add the respective resources-classes to PuPHPeteer. For this we need:

1. Check the release notes for Puppeteer of newly introduced APIs / functionality and create a list
2. Add the relevant classes.

Is this something you think you could help with @xtrime-ru? Of course, we are happy to help along the way and support as needed - don't hesitate to ask questions!

Cheers,
Peter

**Comment by xtrime-ru** [comment]
Hi! We using this wrapper to work with your lib: https://github.com/zoonru/php-puppeteer/tree/master/src/PhpPupeeteer

Probably we need to migrate features from it to your package.

This is what we changed in it on php side to migrate to 5.2 : https://github.com/aktuba/php-puppeteer/compare/master...zoonru:master

I already checked release notes in puppeteer, and there is no significant changes from 1.x version to 5.2. In 5.3 waitFor is deprecated in favour of waitForSelector, waitForTimeout and few other specific methods.

Currently we use this in production with 0 issues.

**Comment by xtrime-ru** [comment]
This is all changes from release notes: 

v1.19.0
class: FileChooser (ea28ccc)
page.waitForFileChooser() (ea28ccc)

v1.20.0
New APIs: element.evaluate, element.evaluateHandle, element.select, jsHandle.evaluate, jsHandle.evaluateHandle

v2.0
New APIs:
page.emulateTimezone(timezoneId)
page.emulateMediaType(type) (superseding the old page.emulateMedia(type) API)
page.emulateMediaFeatures(features)
API changes

New API: page.emulateTimezone(timezoneId) (#4949)
New API: page.emulateMediaType(type) (#5012)
Deprecated API: page.emulateMedia(type) is now obsolete. Use page.emulateMediaType(type) instead. (#5012)
New API: page.emulateMediaFeatures(features) (#5012)

v3.0
Node.js v8.x.x is no longer supported. (#5136)

v3.0.4
Puppeteer should install and run just fine on Node.js v14.1.0 and beyond.

v3.2
page.emulateVisionDeficiency

v5.0
Removal of page.emulateMedia (#6084) This method was deprecated and is now removed from Puppeteer. Swap to page.emulateMediaType which behaves identically.

v5.1
The new Mouse#wheel API: https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#mousewheeloptions

v5.3
New API: page.waitForTimeout and subsequently, frame.waitForTimeout (#6268)
Deprecated API: page.waitFor and frame.waitFor 

v5.3.1
New API: page.emulateIdleState(...) (#6410)

**Comment by spekulatius** [comment]
Hey @xtrime-ru 

Thank you for looking up the API changes - looks like it's a manageable list. 

@nesk and myself are working on improving the test base atm (in preparation for the Puppeteer upgrade). @nesk, which of these APIs would you test for? 

Regarding the PHP Puppeteer fork: It looks like there aren't any larger changes made from zoonru:master back to php-puppeteer:master. But how is it looking back to PuPHPeteer? Unfortunately the git history isn't there anymore. 

Cheers,
Peter

**Comment by nesk** [comment]
Seems like upgrading to the latest version of Puppeteer comes with minimal risks, which is some great news! Thank you @xtrime-ru for your feedback on the upgrade.

@spekulatius Let's take a week to think about missing tests we could do before merging this PR. If we have nothing, let's just merge and hope we will add new tests in the future.

@xtrime-ru Could you rebase your commit and avoid using a specific version in the `package.json` file? I think we can use `~5.0.0`, which will allow the users to use the version they want.

Meanwhile, I will work on the failing tests.

**Comment by spekulatius** [comment]
Hey @nesk 

just started looking into more tests. Opened #115 in this regards - avoiding any issues which probably won't matter anymore. Regarding tests, yes, if we don't have anything more ready soon let's merge it.

Cheers,
Peter

**Comment by xtrime-ru** [comment]
> Regarding the PHP Puppeteer fork: It looks like there aren't any larger changes made from zoonru:master back to php-puppeteer:master. But how is it looking back to PuPHPeteer? Unfortunately the git history isn't there anymore.

As i mentioned https://github.com/zoonru/php-puppeteer not fork, its library which requires puphpeteer via composer, then extends its classes to add phpDocs for IDE's and add some trivial actions. I think its possible to migrate some of this to puphpeteer.

**Comment by spekulatius** [comment]
> > Regarding the PHP Puppeteer fork: It looks like there aren't any larger changes made from zoonru:master back to php-puppeteer:master. But how is it looking back to PuPHPeteer? Unfortunately the git history isn't there anymore.
> 
> As i mentioned https://github.com/zoonru/php-puppeteer not fork, its library which requires puphpeteer via composer, then extends its classes to add phpDocs for IDE's and add some trivial actions. I think its possible to migrate some of this to puphpeteer.

Ah sorry, I missed that part. Definitely worthy to look what could be migrated. There is also some work done on this in the 2.0 branch.

**Comment by xtrime-ru** [comment]
> @xtrime-ru Could you rebase your commit and avoid using a specific version in the `package.json` file? I think we can use `~5.0.0`, which will allow the users to use the version they want.
 
@nesk Will do it later today, after i fixate usage of 5.2.1 in our dependencies.

> Meanwhile, I will work on the failing tests.

Please see this fix in our wrapper lib: https://github.com/aktuba/php-puppeteer/compare/master...zoonru:master#diff-46706d551abfe74b676cb6ab7ebd7fdca8b6a5c466d91a981693c7e1c9e71100R27
I think you will need to do something similar in rialto-php/puphpeteer

**Comment by nesk** [comment]
>> @xtrime-ru Could you rebase your commit and avoid using a specific version in the `package.json` file? I think we can use `~5.0.0`, which will allow the users to use the version they want.
> 
>@nesk Will do it later today, after i fixate usage of 5.2.1 in our dependencies.

@xtrime-ru Actually, I meant `~5.3.0` instead of `~5.0.0`, sorry.

**Comment by nesk** [comment]
Also, could you target the `2.0.0` branch please?

**Comment by xtrime-ru** [comment]
> @xtrime-ru Actually, I meant `~5.3.0` instead of `~5.0.0`, sorry.
Done. Hope i can set puppeteer back to 5.2.1 in our project, since its latest version supported by browserless docker image.

> Also, could you target the `2.0.0` branch please?
Done.

**Comment by xtrime-ru** [comment]
> @xtrime-ru Actually, I meant `~5.3.0` instead of `~5.0.0`, sorry.

@nesk Currently i ended up with ^5.0.0, so we can use 5.2.1 in our projects without issues.

**Comment by xtrime-ru** [comment]
I pushed mots of our internal updates. Including binding HttpRequest to Request class. Now this branch is IDE friendly :) 

Only thing left - is custom Page methods: https://github.com/zoonru/php-puppeteer/blob/master/src/PhpPupeeteer/Resources/Page.php 

But useful thing from there can be added later in separate PR.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@nesk/puphpeteer",
-  "version": "1.6.0",
+  "version": "2.0.0",
```
Comment (nesk) [inline package.json:3]: Do not bother with this, we will do the new release ourselves.

File: package.json [hunk 2]
```diff
@@ -19,10 +19,10 @@
   "license": "MIT",
   "repository": "github:nesk/puphpeteer",
   "engines": {
-    "node": ">=8.0.0"
+    "node": ">=9.0.0"
```
Comment (nesk) [inline package.json:22]: Why is this required?

File: src/PuppeteerProcessDelegate.php [hunk 3]
```diff
@@ -14,6 +14,12 @@ class PuppeteerProcessDelegate implements ShouldHandleProcessDelegation
      */
     public function resourceFromOriginalClassName(string $className): ?string
     {
+        if ($className === 'HTTPResponse') {
+            $className = 'Response';
+        } elseif ($className === 'HTTPRequest') {
+            $className = 'Request';
+        }
+
```
Comment (nesk) [inline src/PuppeteerProcessDelegate.php:22]: This a smart bugfix, but the real fix should be to rename the basic PHP resources:

- [`Request`](https://github.com/rialto-php/puphpeteer/blob/f9a9c17d62076e5e5652df38d38fe26fc565b6f8/src/Resources/Request.php) => `HTTPRequest`
- [`Response`](https://github.com/rialto-php/puphpeteer/blob/f9a9c17d62076e5e5652df38d38fe26fc565b6f8/src/Resources/Response.php) => `HTTPResponse`

**Review by nesk — CHANGES_REQUESTED** [review]
This PR should be dedicated to the version upgrade of Puppeteer. Reviewing so much changes is a real pain 😕

If you want to add some other features, I'm really glad about that! 🙂 But you really should push them in multiple other PRs, so we can review them properly. Please, if you want to upgrade to the latest version of Puppeteer, limit your code changes to that.

However here are some somments about the version upgrade part. 😉

File: src/PuppeteerProcessDelegate.php (hunk 3 above)
Comment (spekulatius) [reply src/PuppeteerProcessDelegate.php:22]: Should we do this separately or as part of this?

Comment (nesk) [reply src/PuppeteerProcessDelegate.php:22]: It should be done in the same PR since its related to (and caused by) the version upgrade. 😉

Comment (spekulatius) [reply src/PuppeteerProcessDelegate.php:22]: okay, makes sense

Comment (xtrime-ru) [reply src/PuppeteerProcessDelegate.php:22]: Yep. https://github.com/puppeteer/puppeteer/releases/tag/v3.2.0
image
Caused by TypeScript issues. So they done this rename. But I like original name more...

If we rename classes, then users will have to fix their code (including us), if they use type checks :(

File: package.json (hunk 2 above)
Comment (xtrime-ru) [reply package.json:22]: https://github.com/puppeteer/puppeteer/releases/tag/v3.0.0
image

**Comment by xtrime-ru** [comment]
> But you really should push them in multiple other PRs, so we can review them properly

@nesk
I agree that this should be moved in separate PRs. I will try to do it on next weekend. But i will be very grateful, if you can review this "as is" and merge, so i don`t need to move commits :) 

Changes in separate commits: 
Снимок экрана 2020-10-26 в 16 08 44

Most of code is just typeHints for IDE. So it should be easy to review.

**Comment by nesk** [comment]
I've fixed the tests, improved some parts of the code and manually merged the code in the branch. Thank you for your contribution @xtrime-ru.

**Comment by xtrime-ru** [comment]
@nesk Glad to hear that. 
May i ask why you choose to use ~5.4.1 instead of ^5.0.0 ?
If there is no requirements in other libs, then they both install latest version. But in our project we need exactly 5.2.1, so npm will install two versions of puppeteer and latest chrome (not the one we need).

**Comment by nesk** [comment]
@xtrime-ru The issue with the `^5.0.0` version constraint, is this part of the code:

https://github.com/rialto-php/puphpeteer/blob/83c906e3073616fe41aee62f96c4d96012d31fca/src/PuppeteerConnectionDelegate.js#L61-L77

Puppeteer doesn't provide any public API to check the class of an object and, in this regard, doesn't provide any garanty that the class names will follow semver. To avoid any breaking issues, I prefer to add a constraint on the minor version instead of the major. And I also prefer to choose the latest version if possible.

If you still want to use a specific version of Puppeteer (I understand you have your own requirements), [you can take a look at NPM shrinkwrap to override the dependencies](https://stackoverflow.com/a/17423915/1513045).

**Comment by xtrime-ru** [comment]
@nesk Thanks for link to shrinkwrap! Probably, this is good alternative for using own fork :) 

You mean, you can't check if HTTPRequest, HTTPResponse and WebWorker classes exists? But they were renamed in 3.3: https://github.com/puppeteer/puppeteer/releases/tag/v3.3.0 , so your code should work with any 5.x version.

**Comment by nesk** [comment]
Yes, the code should work, but what if the `5.5` version changes some names again? [It already happened once in the past](https://github.com/puppeteer/puppeteer/issues/3067), so I prefer to avoid any issues.

One day [we will provide the ability to support all Puppeteer versions](https://github.com/rialto-php/puphpeteer/issues/23), but for moment what have to stick to only a few versions.

**Comment by xtrime-ru** [comment]
Good point. But how about `>=3.3 <=5.4` or something similar? 
PS. Found this handy semver checker: https://semver.npmjs.com

**Comment by nesk** [comment]
Because this is something I want to handle in #23, I feel like we should be careful with this, so I don't want to rush things. The latest version of Puppeteer will be just fine for the majority, and if some users really want a specific Puppeteer version, then they can use shrinkwrap. 😉
